sdks/swift/Sources/ApacheBeam/Runners/PortableRunner.swift

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Logging
+import GRPC
+import NIOCore
+
+public struct PortableRunner : PipelineRunner {
+    let loopback:Bool
+    let log: Logging.Logger
+    let host: String
+    let port: Int
+    
+    public init(host:String="localhost",port:Int=8073,loopback:Bool=false) {
+        self.loopback = loopback
+        self.log = .init(label: "PortableRunner")
+        self.host = host
+        self.port = port
+    }
+    
+    public func run(_ context: PipelineContext) async throws {
+        var proto = context.proto
+        if loopback {
+            //If we are in loopback mode we want to replace the default environment
+            //with an external environment that points to our local worker server
+            let worker = try WorkerServer(context.collections,context.pardoFns)
+            log.info("Running in LOOPBACK mode with a worker server at \(worker.endpoint).")
+            proto.components.environments[context.defaultEnvironmentId] = try .with {
+                try Environment.external(worker.endpoint).populate(&$0)
+            }
+        }
+        log.info("\(proto)")
+        log.info("Connecting to Portable Runner at \(host):\(port).")
+        let group = PlatformSupport.makeEventLoopGroup(loopCount: 1)
+        let channel = try GRPCChannelPool.with(target: .host(host, port: port),
+                                           transportSecurity: .plaintext, eventLoopGroup: group)
+        let client = Org_Apache_Beam_Model_JobManagement_V1_JobServiceAsyncClient(channel: channel)
+        let prepared = try await client.prepare(.with {
+            $0.pipeline = proto
+        })
+        let job = try await client.run(.with {
+            $0.preparationID = prepared.preparationID
+        })
+        log.info("Submitted job \(job.jobID)")
+        var done = false
+        while !done {
+            let status = try await client.getState(.with {
```

Comment:
I think where prism is failing with Swift is that I've currently got the code to be a bit blocking on state & message transitions, and I'm not sure I have it set up to "do the right thing" with the unary getState calls, vs getStateStream or getMessageStream, so eventually it gets blocked up.

Not a 100% explanation for what we've seen, but better than nothing.

Great info! I'll let you know when it's resolved.

Reply:
Back to stumped after replicated this code. Prism is happy to run this approach from the Go SDK side. I might try seeing how hard it is to get the Swift SDK running in Linux, and answer my questions directly.

Reply:
I tested on Linux x86 but haven't tried Linux ARM yet. In theory when I was messing around with timestamps I'm doing the right thing with endianness there but I haven't verified that everywhere yet. In any case, "swift test" should work with a beta 5.9 build of Swift. I'm currently moving some of that to an "Examples" folder to make this sort of thing easier. I mostly used tests initially because it's a one button command in my IDE.

Reply:
I've successfully installed Swift (5.9 dev snap) on my personal laptop and pulled down the PR locally, and can replicate the behaviour on prism that you were seeing. I should be able to get to the bottom of those problems, and make prism more robust!

Hot dang do I love the clarity in the Swift Docs on installing on different OS flavours, and how automated `swift test` is. It failed the first time, and clearly said what else I needed to install (openSSL) to clear it up.

We should also add the following to the beam repo's .gitignore, right under the python section.

```
# Ignore files generated by the Swift build process
sdks/swift/.build/**/*
```

Building the swift SDK lead to many many files in there.

Reply:
The mystery problem has gone deeper. Prism is hosting worker services at X, swift is trying to connect to X, and then prism apparently never gets that connection.

Reply:
Interesting. The worker code is connecting to Prism on the port specified so that doesn't appear to be a problem, though it's a bit hard due to the "backwards" nature of the control plane connection. And, yes, the Swift toolchain is surprisingly pleasant to use. With the plugin system we should also be able to add some nice quality-of-life Beam features to SPM as well.

Reply:
Not sure what's going on. Prism does the same thing with both the jobmanagement port and the worker port. I added a debugz page with thread dumps and it just shows prism is waiting on a worker to connect, so it can produce work.

In the short term I'll try and get prism to gracefully fail the job if it doesn't get a control connection in a minute.

Reply:
Note for tomorrow: the tests don't fail if the pipeline fails.

Reply:
Got it!

I just let Go's net listener determine the endpoint host spec. For local host, it uses [::]:PORT, which it's fine with parsing itself when it receives it.

The bit that made it tricky is I didn't notice the GRPCChannel override to support the beam api endpoint descriptor. But when I did, i hard changed the host to "localhost" and then the pipeline fails with a reasonable "whats's a split request?" error.

Prisms at fault here, and has become more resilient as a result. #winning

Reply:
Ah, excellent. Is the problem that Go is using IPv6 and I'm connecting on IPv4? Most of the code I looked at seemed to be doing the localhost thing so I just sort of copied that... but that's probably suboptimal in the year 2023 (At some point I also need to add the TLS support too)

Reply:
Here's the sequence of events:

I ask for a port ":0", to listen on on any IP this machine is on.

It gives me "[::]:PORT"

but [::] is ipv6 unspecified.

In Go, that's explicitly handed as the loopback address.

The fix on my end is to just say the endpoint is one of "localhost:PORT" or "[::1]:PORT".

The next failure does happen on Swift side (KV<String, String> isn't KV<String,Optional<String>>) which either means prism messed up coders, or the swift SDK did, and this is just revealing something due to prism executing a single transform.

I need to make prism gracefully mark the job failed on that SDK worker failure though.

Reply:
```
2023-08-28T09:35:31-0700 info Source : [ApacheBeam] Waiting for input on inst003-stage004_source
Could not cast value of type 'ApacheBeam.KV<Swift.String, Swift.Optional<Swift.String>>' (0x7f3a980093f8) to 'ApacheBeam.KV<Swift.String, Swift.String>' (0x7f3adf6cf438).

```